django/db/models/query.py

```diff
@@ -942,6 +942,8 @@ def extra(self, select=None, where=None, params=None, tables=None,
 
     def reverse(self):
         """Reverse the ordering of the QuerySet."""
+        if not self.query.can_filter():
+            raise TypeError('Cannot reverse a query once a slice has been taken')
```

Comment:
I think that `AssertionError` is the most common in the `Query`. It's also used with the same check in other methods e.g. `Query.extra` and `Query._earliest_or_latest`:
```python
assert self.query.can_filter(), \
    "Cannot change a query once a slice has been taken."
```
therefore it will probably be better to keep consistency.